File: src/Phan_01/encapsulation_accessModifier.cpp
#include<iostream>
#include<string>
using namespace std;

class Animal {
private:
	string name;
	string description() {
		return name + " is " + to_string(age) + " years old\n";
	}
protected:
	int age;
	string run(int distance) {
		return name + " run " + to_string(distance) + " miles\n";
	}
public:
	string species;
	string speak(string sound) {
		return name + " says " + sound + "\n";
	}

	Animal(string name, int age, string species) {
		this->name = name;
		this->age = age;
		this->species = species;
	}
	// getter methods
	string get_name() {
		return this->name;
	}
	// setter methods
	void set_name(string name) {
		this->name = name;
	}
};

int main() {
	Animal animal("rocky", 10, "dog");

	// cout << "name: " << animal.name << endl;         // error
	// cout << "age: " << animal.age << endl;           // error
	cout << "species: " << animal.species << endl;      // success

	// cout << animal.description() << endl;            // error
	// cout << animal.run(10) << endl;                  // error
	cout << animal.speak("Grr..") << endl;              // success

	cout << animal.get_name() << endl;	// rocky
	animal.set_name("kuro");
	cout << animal.get_name() << endl; // kuro

	return 0;
}

File: src/Phan_01/encapsulation_accessModifier.py
class Animal:
    def __init__(self, name, age, species):
        # private attributes
        self.__name = name

        # protected attributes
        self._age = age

        # public attributes
        self.species = species

    # private methods
    def __description(self):
        return f"{self.__name} is {self._age} years old"

    # protected methods
    def _run(self, distance):
        return f"{self.__name} run {distance} miles"

    # public methods
    def speak(self, sound):
        return f"{self.__name} says {sound}"

    # getter methods
    def get_name(self):
        return self.__name

    # setter methods
    def set_name(self, name):
        self.__name = name

    # property functions
    name = property(get_name, set_name)

    # using property decorator
    # getter methods
    @property
    def name(self):
        return self.__name

    # setter methods
    @name.setter
    def name(self, name):
        self.__name = name


animal = Animal("rocky", 10, "dog")

# check attributes access modifier
# print("name:", animal.__name)       # error
# print("age: ", animal._age)         # error
print("species:", animal.species)  # success

# check methods access modifier
# print(animal.__description())     # error
# print(animal._run(10))            # error
print(animal.speak("Grr.."))  # success

print(animal.get_name())  # rocky
animal.set_name("kuro")
print(animal.get_name())  # kuro

print(animal.name)  # kuro
animal.name = "yasuo"
print(animal.name)  # yasuo
